src/ophys_etl/modules/segmentation/merge/roi_merging.py

```diff
@@ -0,0 +1,378 @@
+            recently_merged.add(new_roi.roi_id)
+            roi_lookup[seed_roi.roi_id] = new_roi
+
+            # all ROIs that neighbored child_roi now neighbor new_roi
+            severed_neighbors = neighbor_lookup.pop(child_roi.roi_id)
+            severed_neighbors = severed_neighbors.intersection(valid_roi_id)
+            for roi_id in severed_neighbors:
+                if roi_id == new_roi.roi_id:
+                    continue
+                neighbor_lookup[new_roi.roi_id].add(roi_id)
+                neighbor_lookup[roi_id].add(new_roi.roi_id)
+
+        # break out any ROIs that are too large
+        k_list = list(roi_lookup.keys())
+        for roi_id in k_list:
+            if roi_lookup[roi_id].area >= anomalous_size:
+                roi = roi_lookup.pop(roi_id)
+                roi.valid_roi = False
+                anomalous_rois[roi_id] = roi
+
+                valid_roi_id.remove(roi_id)
+                neighbor_lookup.pop(roi_id)
+                sub_video_lookup.pop(roi_id)
+                timeseries_lookup.pop(roi_id)
+
+        # remove an obsolete ROI IDs from neighbor lookup
+        for roi_id in neighbor_lookup:
+            new_set = neighbor_lookup[roi_id].intersection(valid_roi_id)
+            neighbor_lookup[roi_id] = new_set
+
+        # remove ROIs that have changed from sub_video and
+        # merger metric lookup tables
+        for roi_id in recently_merged:
+            if roi_id in sub_video_lookup:
+                sub_video_lookup.pop(roi_id)
+
+        # remove non-existent ROIs and ROIs whose areas
+        # have significantly changed from timeseries_lookup
+        # (calculating timeseries_lookup is expensive, so
+        # we allow ROIs to grow a little before recalculating)
+        k_list = list(timeseries_lookup.keys())
+        for roi_id in k_list:
+            pop_it = False
+            area0 = timeseries_lookup[roi_id]['area']
+            if roi_id not in valid_roi_id:
+                pop_it = True
+            elif roi_lookup[roi_id].area > 1.05*area0:
+                pop_it = True
+
+            if pop_it:
+                timeseries_lookup.pop(roi_id)
+
+        # remove all recently merged ROIs from the merger metric
+        # lookup table
+        merger_keys = list(merger_to_metric.keys())
+        for pair in merger_keys:
+            if pair[0] in recently_merged or pair[1] in recently_merged:
+                merger_to_metric.pop(pair)
+
+        logger.info(f'merged {n0} ROIs to {len(roi_lookup)}; '
+                    f'{len(anomalous_rois)} anomalous ROIs; '
+                    f'after {time.time()-t0:.2f} seconds')
+
+        # make sure we did not lose track of any ROIs
+        for roi_id in incoming_rois:
+            if roi_id not in have_been_merged:
+                if roi_id not in roi_lookup:
+                    if roi_id not in anomalous_rois:
+                        raise RuntimeError(f"lost track of {roi_id}")
```

Comment:
is this a possibility?
if this was a check for development purposes, it seems like this should be covered in a test.

Reply:
It should not be a possibility. Or, put another way, if this triggers, it means something is wrong with the algorithm.

However, I would like to leave it in there for now. While I've tried to add unit tests for every function in the module, I can't guarantee that the tests exercise all edge cases. I understand why this is not considered best practice, but I feel more comfortable knowing that check is there, at least in this early development stage.

Reply:
ok